### DISCUSSION TITLE
Trouble setting up ESP8266 controller

### DISCUSSION THREAD

**Discussion by nborof** [thorrak/fermentrack#642] [Q&A]
Hi there -

I've been trying to tentatively expand my use of Fermentrack to function as a controller.  I bought one of these ESP8266 modules:
https://www.amazon.com/gp/product/B081CSJV2V

But I've been having trouble setting it up.  I managed to flash it with the most current BrewPi firmware, and I got it to connect to the network (where it can be successfully pinged).  But when I set it up, it doesn't seem to connect to Fermentrack entirely.  When I try to Debug Device Connection, it passes all the tests except for the Connection Test. 
![Screen Shot 2021-05-26 at 4 12 28 PM](https://user-images.githubusercontent.com/38414932/119748711-2101ce80-be5b-11eb-8fae-31bfce85477b.png)

I also tried flashing it as a serial connection.  When I did this, everything except the USB Serial Number Test passed during the debugging.  However, when I went to Configure Sensors/Pins, I was unable to set a pin (there were three pins available) for either the chamber or the room (outside) temp.  I could set a pin for the heating or cooling relays, but when I tried to configure a thermometer, it said, "Failed to write the configuration to the controller. If this continues, try resetting the EEPROM on the controller." I tried resetting the EEPROM, and this did not help.

So my questions is, is this particular ESP8266 not really compatible?  They were the first search result on Amazon for "ESP8266", so I figured it was a reasonably safe bet.  Since I bought the multi-pack, I did test all of this with two different units, and had the same results each time, so I don't think it's just a single defective unit.

**Comment by thorrak** [comment]
Although Fermentrack doesn't enforce the restrictions for historical reasons, pins can only be used for heating/cooling relays & door sensors, and OneWire temperature sensors must be used for chamber ("fridge"), beer, and room sensors. Additionally, OneWire sensors must have their data pin connected to pin D6. 

Although the board I recommend is the [Lolin D1 Mini](https://www.aliexpress.com/item/32529101036.html), (especially in conjunction with one of my [PCB designs](https://github.com/thorrak/thorrak_hardware/tree/master/ESP8266%20BrewPi%20Boards)) the boards you linked are "NodeMCU" boards and should work just fine - both for Serial and WiFi - as long as you wire everything to the correct pins.

As far as the issues you are having with WiFi are concerned, what happens if you telnet to 192.168.4.150 and type a single "l" and then hit enter? You should get something like `L:["Mode   Fridge Const.","Beer   70.1  --.- �F","Fridge 68.6  28.4 �F","Wait to Cool   09m28"]` back (the text that would be displayed on the LCD screen of your controller, if you had one connected).

**Reply by nborof** [reply]
When I try to connect via telnet, I get 

Trying 192.168.4.150...
telnet: Unable to connect to remote host: No route to host

Also, I've noticed the boards seem to have some difficulty with authenticating on the wireless network.  They seem to take several attempts (re-entering the credentials each time) before they finally connect.  When I turned off the wifi password on the router, they connected right away (but they still had the same problems as before: Fermentrack can't communicate, telnet unable to connect).  I'm guessing these boards are either somehow incompatible with BrewPi or I just got a bad bunch.  

I'm thinking I might try another kind of board. The Lolin D1 Mini you linked to will take a month or two to get to me from AliExpress, so I was curious if most D1 Mini boards are basically the same? For example:

https://www.amazon.com/Organizer-ESP8266-Internet-Development-Compatible/dp/B081PX9YFV
https://www.amazon.com/Aceirmc-ESP8266-Internet-Development-Compatible/dp/B07V84VWSM

**Reply by lbussy** [reply]
I've had good luck with the clones, yes. Might as well give it a try. Your experience doesn't match my own with this project, so it might be the board.

**Reply by thorrak** [reply]
Separately, I would recommend trying the "BrewPi-ESP8266 (DO NOT USE) - LowDelay" firmware just to see if it does a better job of connecting/staying connecting as that firmware has more recent WiFi libraries. If that helps, I can compile some normal (ie. not 'lowdelay') firmware for you that you can use.

**A word of caution** -- Do not use this firmware while your controller is connected to a relay. This firmware doesn't have the compressor protections that normal BrewPi firmware has, and may destroy whatever devices you have connected. Give it a test, see if it works, and if it does I'll get you a version that has all the normal protections enabled.

**Reply by nborof** [reply]
I gave the "DO NOT USE" firmware a try, and while it seemed to authenticate with the router faster, the same problems were there - fermentrack and telnet were unable to connect to the board.

**Reply by nborof** [reply]
I ordered one of the D1 clones.  Hopefully that will do the trick. :-)

**Reply by nborof** [reply]
I got the D1 clone, and the problems seem to be identical.  I tried using an older router with just the raspberry pi, the D1, and a laptop connected to it (and no internet access), and I discovered something that makes very little sense to me.  I can get fermentrack to discover and connect with the ESP8266 if the subnet is 192.168.0.0, 192.168.1.0, 192.168.2.0, or 192.168.3.0.  But if the subnet is 192.168.4.0 (which of course is what my home network uses), fermentrack cannot discover the ESP8266 controller.  Again, I tested this on an isolated router that was not connected to the internet or any other network with only three devices.  The only think I changed was the router's ip from 192.168.1.1 (e.g.) to 192.168.4.1.

Could this be a bug in fermentrack?  Or is there something about mDNS searches using the 192.18.4.0 subnet that I don't understand? 

So curious!

**Reply by nborof** [reply]
The only think I can think of is that the internal network of the BrewPi firmware seems to be on that subnet (i.e., you go to 192.168.4.1 to set up the initial wifi).  Is there a way to change the default subnet in the firmware (or elsewhere)?  Obviously, I can change the subnet for my home router, but it requires a bit of work in my situation, and I'd rather not go through that if there's a simpler fix. :-)

**Reply by lbussy** [reply]
It sounds like WiFiManager is not fully disconnecting/disposing, causing a routing issue.

**Reply by nborof** [reply]
That would make sense. I'm guessing that's not a simple fix, right? It's not a huge deal to change the addresses on my home network, so that's probably the best way for me to go.  Hopefully this will help anyone else who runs into a similar problem.

**Reply by thorrak** [reply]
I just uploaded a set of firmware that has updated libraries. It might solve this issue if you want to give it a try. It's available in Fermentrack/BrewFlasher as BrewPi-ESP8266 (Beta) v. 0.13.

**Reply by nborof** [reply]
I tried it, and Fermentrack wasn't able to discover the BrewPi, even when they were on a different subnet (.3). I re-flashed to Version: 2.4.2, Revision: 0.11, and it works fine again (just not if the network is on the .4 subnet).

**Reply by thorrak** [reply]
One thing to note (not sure if it applies here) - mDNS responses can't cross subnets. While Fermentrack should absolutely be able to reach/locate devices located on a different subnet than it is on (assuming routing is set up) it won't be able to autodetect devices set up this way.

**Comment by vanosg** [comment]
I'm confused as to how you can ping it though- it shouldn't be able to be pinged if it's on .4 subnet unless you have some unexpected routing already added. If you can ping it, you should be able to telnet to it. And of course I'm not there to know what steps you have and haven't done so we enter pure spectulation, but it sounds like you didn't connect to the ESP directly via wifi, as mentioned in https://github.com/thorrak/fermentrack/wiki . Just a thought, hope it helps

**Reply by nborof** [reply]
I am confused by this as well.  The pinging on the .4 subnet was inconsistent - sometimes it would work, and other times it would not.  The telnet however, never worked on the .4 subnet, and Fermentrack was never able to connect successfully.

**Reply by lbussy** [reply]
Ping is ICMP and stateless, Telnet is TCP and stateful.  If the AP does not go away, two different 192.168.4.0/24 networks will be alive on the ESP and while sometimes it will connect, sometimes it will not.  It makes sense that the ping works "enough" to get returns (albeit inconsistently), and mDNS is UDP (also stateless) and could sometimes resolve depending on the client.
